File: crawler/src/processors/db_writer.ts
import * as crypto from "crypto";
import prisma from "../db";
import logger from "../logger";
import { cleanContent, truncateSummary } from "./content_cleaner";
import { checkSensitive } from "./sensitive_checker";
import { computeSimHash } from "../engine/deduper";
import { EnrichedQuote, RawQuote } from "../providers/base_provider";

function md5(text: string): string {
  return crypto.createHash("md5").update(text, "utf-8").digest("hex");
}

function normalizeForDedup(text: string): string {
  return text.replace(/[\s，。！？、；：""''（）《》【】]/g, "").toLowerCase();
}

export interface WriteResult {
  quoteId: number | null;
  crawlRecordId: number | null;
  status: "inserted" | "skipped_duplicate" | "skipped_sensitive" | "error";
  reason?: string;
}

export async function writeQuote(enriched: EnrichedQuote, sourceId?: number, taskId?: number): Promise<WriteResult> {
  const cleanedContent = cleanContent(enriched.content);
  if (!cleanedContent || cleanedContent.length < 3) {
    return { quoteId: null, crawlRecordId: null, status: "error", reason: "内容过短或为空" };
  }

  const contentMd5 = md5(normalizeForDedup(cleanedContent));
  const simHash = enriched.simHash || computeSimHash(cleanedContent);

  const sensitiveResult = checkSensitive(cleanedContent);
  const auditStatus = sensitiveResult.hasSensitive ? 2 : 1;

  // CrawlRecord 临时库去重（MD5 指纹）
  const existingRecord = await prisma.crawlRecord.findFirst({
    where: { fingerprint: contentMd5 },
  });
  if (existingRecord) {
    logger.debug(`CrawlRecord 指纹重复跳过: ${contentMd5}`);
    return {
      quoteId: existingRecord.finalQuoteId,
      crawlRecordId: existingRecord.id,
      status: "skipped_duplicate",
      reason: "CrawlRecord 指纹已存在",
    };
  }

  // 正式 Quote 表去重（MD5）
  const existingQuote = await prisma.quote.findFirst({
    where: { contentMd5 },
  });
  if (existingQuote) {
    logger.debug(`Quote 表 MD5 重复跳过: ${cleanedContent.slice(0, 20)}...`);
    return {
      quoteId: existingQuote.id,
      crawlRecordId: null,
      status: "skipped_duplicate",
      reason: "Quote 表已存在（MD5）",
    };
  }

  const crawlRecord = await prisma.crawlRecord.create({
    data: {
      taskId: taskId || 0,
      sourceId: sourceId || 0,
      originalUrl: enriched.url,
      content: cleanedContent,
      author: enriched.author,
      sourceName: enriched.source,
      rawHtml: null,
      extractMethod: "provider",
      simHash,
      fingerprint: contentMd5,
      categoryType: enriched.categoryType,
      categoryIds: enriched.tags ? enriched.tags.join(",") : null,
      auditStatus,
      auditReason: sensitiveResult.hasSensitive
        ? `命中敏感词: ${sensitiveResult.matchedWords.join(",")}`
        : null,
      auditBy: "crawler",
      auditAt: new Date(),
      finalQuoteId: null,
    },
  });

  if (sensitiveResult.hasSensitive) {
    logger.warn(`敏感内容已标记 auditStatus=2: ${sensitiveResult.matchedWords.join(",")}`);
    return {
      quoteId: null,
      crawlRecordId: crawlRecord.id,
      status: "skipped_sensitive",
      reason: `命中敏感词: ${sensitiveResult.matchedWords.join(",")}`,
    };
  }

  const summary = truncateSummary(cleanedContent, 100);

  const quote = await prisma.quote.create({
    data: {
      content: cleanedContent,
      contentMd5,
      simHash,
      author: enriched.author || null,
      source: enriched.source || null,
      summary,
      tags: enriched.tags ? enriched.tags.join(",") : null,
      wordCount: cleanedContent.length,
      isFree: true,
      isActive: true,
      quoteType: "crawl",
    },
  });

  await prisma.crawlRecord.update({
    where: { id: crawlRecord.id },
    data: { finalQuoteId: quote.id },
  });

  logger.info(`入库成功 Quote#${quote.id}: ${cleanedContent.slice(0, 30)}...`);

  return {
    quoteId: quote.id,
    crawlRecordId: crawlRecord.id,
    status: "inserted",
  };
}

export async function batchWrite(enrichedList: EnrichedQuote[], sourceId?: number, taskId?: number): Promise<WriteResult[]> {
  const results: WriteResult[] = [];
  for (const item of enrichedList) {
    try {
      const result = await writeQuote(item, sourceId, taskId);
      results.push(result);
    } catch (err: any) {
      logger.error(`入库异常: ${err.message}`);
      results.push({ quoteId: null, crawlRecordId: null, status: "error", reason: err.message });
    }
  }
  return results;
}

export default { writeQuote, batchWrite };

File: crawler-py/crawler/processors/db_writer.py
"""入库写入

对应 TypeScript 版 crawler/src/processors/db_writer.ts。
去重（复用 server-py utils/dedup.py 的 MD5/SimHash）→ 写 CrawlRecord → 写 Quote。
重复内容更新 CrawlRecord.auditStatus=3 后跳过。
"""
from datetime import datetime

from ..db import CrawlRecord, Quote, get_session
from ..engine.deduper import compute_sim_hash, dedup
from ..logger import logger
from .content_cleaner import clean_content, truncate_summary
from .sensitive_checker import check_sensitive


def normalize_for_dedup(text: str) -> str:
    """对应 TS normalizeForDedup（复用 dedup.normalize）"""
    return dedup["normalize"](text)


def md5(text: str) -> str:
    """对应 TS md5（复用 dedup.md5）"""
    return dedup["md5"](text)


def write_quote(enriched: dict, source_id: int = None, task_id: int = None) -> dict:
    """对应 TS writeQuote

    enriched: EnrichedQuote 字典（content/author/source/url/tags/categoryType/simHash...）
    返回 {quoteId, crawlRecordId, status, reason}
    """
    cleaned_content = clean_content(enriched.get("content") or "")
    if not cleaned_content or len(cleaned_content) < 3:
        return {"quoteId": None, "crawlRecordId": None, "status": "error", "reason": "内容过短或为空"}

    # 复用 server-py dedup：MD5 指纹 + SimHash
    fp = dedup["fingerprint"](cleaned_content)
    content_md5 = fp["contentMd5"]
    sim_hash = enriched.get("simHash") or compute_sim_hash(cleaned_content)

    sensitive_result = check_sensitive(cleaned_content)
    audit_status = 2 if sensitive_result["hasSensitive"] else 1

    db = get_session()
    try:
        # CrawlRecord 临时库去重（MD5 指纹）
        existing_record = (
            db.query(CrawlRecord).filter(CrawlRecord.fingerprint == content_md5).first()
        )
        if existing_record:
            # 重复：更新 auditStatus=3 后跳过
            existing_record.auditStatus = 3
            existing_record.auditReason = "CrawlRecord 指纹已存在"
            db.commit()
            logger.debug(f"CrawlRecord 指纹重复跳过: {content_md5}")
            return {
                "quoteId": existing_record.finalQuoteId,
                "crawlRecordId": existing_record.id,
                "status": "skipped_duplicate",
                "reason": "CrawlRecord 指纹已存在",
            }

        # 正式 Quote 表去重（MD5）
        existing_quote = db.query(Quote).filter(Quote.contentMd5 == content_md5).first()
        if existing_quote:
            logger.debug(f"Quote 表 MD5 重复跳过: {cleaned_content[:20]}...")
            return {
                "quoteId": existing_quote.id,
                "crawlRecordId": None,
                "status": "skipped_duplicate",
                "reason": "Quote 表已存在（MD5）",
            }

        tags = enriched.get("tags")
        crawl_record = CrawlRecord(
            taskId=task_id or 0,
            sourceId=source_id or 0,
            originalUrl=enriched.get("url") or "",
            content=cleaned_content,
            author=enriched.get("author"),
            sourceName=enriched.get("source"),
            rawHtml=None,
            extractMethod="provider",
            simHash=sim_hash,
            fingerprint=content_md5,
            categoryType=enriched.get("categoryType"),
            categoryIds=",".join(tags) if tags else None,
            auditStatus=audit_status,
            auditReason=(
                f"命中敏感词: {','.join(sensitive_result['matchedWords'])}"
                if sensitive_result["hasSensitive"] else None
            ),
            auditBy="crawler",
            auditAt=datetime.now(),
            finalQuoteId=None,
        )
        db.add(crawl_record)
        db.commit()
        db.refresh(crawl_record)

        if sensitive_result["hasSensitive"]:
            logger.warning(f"敏感内容已标记 auditStatus=2: {','.join(sensitive_result['matchedWords'])}")
            return {
                "quoteId": None,
                "crawlRecordId": crawl_record.id,
                "status": "skipped_sensitive",
                "reason": f"命中敏感词: {','.join(sensitive_result['matchedWords'])}",
            }

        summary = truncate_summary(cleaned_content, 100)
        quote = Quote(
            content=cleaned_content,
            contentMd5=content_md5,
            simHash=sim_hash,
            author=enriched.get("author") or None,
            source=enriched.get("source") or None,
            summary=summary,
            tags=",".join(tags) if tags else None,
            wordCount=len(cleaned_content),
            isFree=True,
            isActive=True,
            quoteType="crawl",
        )
        db.add(quote)
        db.commit()
        db.refresh(quote)

        crawl_record.finalQuoteId = quote.id
        db.commit()

        logger.info(f"入库成功 Quote#{quote.id}: {cleaned_content[:30]}...")

        return {
            "quoteId": quote.id,
            "crawlRecordId": crawl_record.id,
            "status": "inserted",
        }
    finally:
        db.close()


def batch_write(enriched_list: list, source_id: int = None, task_id: int = None) -> list:
    """对应 TS batchWrite"""
    results = []
    for item in enriched_list:
        try:
            result = write_quote(item, source_id, task_id)
            results.append(result)
        except Exception as err:
            logger.error(f"入库异常: {err}")
            results.append({"quoteId": None, "crawlRecordId": None, "status": "error", "reason": str(err)})
    return results
